### ISSUE TITLE
When saving query, front end uses non-distinct array agg function.

### ISSUE THREAD

**Issue by mathemancer** [centerofci/mathesar#2334]
## Description

When creating a query using the Data Explorer, one can summarize columns, aggregating to a list. The list only has one of each member. This is desirable in most cases. However, the request sent to _save_ the query does not use the same aggregation. It uses the non-distinct version.

- The non-distinct version is generally less desirable.
- This is confusing, since then the version of the query saved is different from what the user expects.
Example of the on-the-fly request:
```
POST http://localhost:8000/api/db/v0/queries/run/
```
```json
{
  "base_table": 503,
  "initial_columns": [
    {
      "alias": "Talks_id",
      "id": 1894
    },
    {
      "alias": "Organizations_Organization",
      "id": 1867,
      "jp_path": [
        [
          1880,
          1892
        ],
        [
          1875,
          1890
        ]
      ]
    },
    {
      "alias": "Topics_Name",
      "id": 1876,
      "jp_path": [
        [
          1894,
          1884
        ],
        [
          1885,
          1896
        ]
      ]
    }
  ],
  "transformations": [
    {
      "type": "hide",
      "spec": [
        "Talks_id"
      ]
    },
    {
      "type": "summarize",
      "spec": {
        "base_grouping_column": "Organizations_Organization",
        "grouping_expressions": [
          {
            "input_alias": "Organizations_Organization",
            "output_alias": "Organizations_Organization_grouped"
          }
        ],
        "aggregation_expressions": []
      }
    }
  ],
  "display_names": {
    "Talks_id": "Talks_id",
    "Organizations_Organization": "Organizations_Organization",
    "Topics_Name": "Topics_Name"
  },
  "parameters": {
    "limit": 100,
    "offset": 0
  }
}
```
Example of the malformed 'save' query:
```
POST http://localhost:8000/api/db/v0/queries/
```
```json
{
  "name": "test query pls ignore",
  "description": "test",
  "base_table": 503,
  "initial_columns": [
    {
      "alias": "Talks_id",
      "id": 1894
    },
    {
      "alias": "Organizations_Organization",
      "id": 1867,
      "jp_path": [
        [
          1880,
          1892
        ],
        [
          1875,
          1890
        ]
      ]
    },
    {
      "alias": "Topics_Name",
      "id": 1876,
      "jp_path": [
        [
          1894,
          1884
        ],
        [
          1885,
          1896
        ]
      ]
    }
  ],
  "transformations": [
    {
      "type": "hide",
      "spec": [
        "Talks_id"
      ]
    },
    {
      "type": "summarize",
      "spec": {
        "base_grouping_column": "Organizations_Organization",
        "grouping_expressions": [
          {
            "input_alias": "Organizations_Organization",
            "output_alias": "Organizations_Organization_grouped"
          }
        ],
        "aggregation_expressions": [
          {
            "input_alias": "Topics_Name",
            "output_alias": "Topics_Name_agged",
            "function": "aggregate_to_array"
          }
        ]
      }
    }
  ],
  "display_names": {
    "Talks_id": "Talks_id",
    "Organizations_Organization": "Organizations_Organization",
    "Topics_Name": "Topics_Name"
  }
}
```

Note the difference at `'.transformations[1].spec.aggregation_expressions[0].function'` (`jq` query style).

## Expected behavior

The saved query should use the distinct version of array aggregation. Generally, saved queries should never be different from the 'on-the-fly' query associated with them. We should consider setting things up to guarantee this.

## To Reproduce

Create a query with an array aggregation which aggregates a non-distinct column. Save the query. Notice the difference.

**Comment by mathemancer** [comment]
@kgodey I put this in usability testing blockers given how jarring the experience was for me, but you may want to re-prioritize as you see fit.

**Comment by pavish** [comment]
@mathemancer 

> Generally, saved queries should never be different from the 'on-the-fly' query associated with them. We should consider setting things up to guarantee this.

This is part of the architecture of auto-summarization. In the 'on-the-fly' query, notice that we do not supply anything in `aggregation_expressions`. This leads to the server calculating where to place the columns and the result provided by the server places
```
{
  "input_alias": "Topics_Name",
  "output_alias": "Topics_Name_agged",
  "function": "aggregate_to_array"
}
```
within `aggregation_expressions`. 

The frontend then stores the calculated query.

> The saved query should use the distinct version of array aggregation

I believed all our array aggregation was distinct. Either way, this is a backend issue. The response returned by the server would be the same as the saved query, which the frontend stores.

**Comment by mathemancer** [comment]
@pavish Okay, good to know; thanks for the walkthrough. While I'm grumpy that the work has to be done in the back end, I'm happy we can make reasonable guarantees about what the front end is going to send to be saved.

**Comment by Subhojit-Dey1234** [comment]
@mathemancer @pavish  Can you explain a little bit? I am not able to understand the issue.

**Comment by mathemancer** [comment]
> This is part of the architecture of auto-summarization. In the 'on-the-fly' query, notice that we do not supply anything in `aggregation_expressions`. This leads to the server calculating where to place the columns and the result provided by the server places
> 
> ```
> {
>   "input_alias": "Topics_Name",
>   "output_alias": "Topics_Name_agged",
>   "function": "aggregate_to_array"
> }
> ```
> 
> within `aggregation_expressions`.
> 
> The frontend then stores the calculated query.

This is inaccurate. I just double-checked the sequence of requests and responses. The back end is returning `distinct_aggregate_to_array`, as it should, when `POST .../queries/run/` is called with the above request (or any that doesn't specify the aggregation function). The front end, for some reason, is changing this to `aggregate_to_array` when it then calls the raw `POST .../queries/` endpoint. This is a front end problem. `aggregate_to_array` and `distinct_aggregate_to_array` are (and should) be distinct functions. 

Moreover, the root problem is that the front end is changing the returned query _at all_ when saving it. It should never do this.

**Comment by mathemancer** [comment]
@Subhojit-Dey1234 We have an endpoint, `POST /api/db/v0/queries/run/`, that lets the front end (or anyone) run a query by submitting an appropriate JSON defining such a query. Such a JSON is given in the first example above. The endpoint fills in some pieces of the query (if needed) and runs it. The response to a request as above includes the fleshed-out version of the query that was run. Separately, the front end (or anyone) can persist a query as a Django model instance by submitting a JSON defining the query to `POST /api/db/v0/queries/`. Such a JSON is given in the second example above. 

The problem (as I see it) is that during a specific user flow where the user is building a query, the front end is showing them results of running that query by hitting the `/api/db/v0/queries/run/` endpoint as appropriate. Then, the user clicks "Save" to persist the query. At that point, the front end is sending a request to save a query with a JSON that defines a different query than that which was returned by the most recent response from the `/api/db/v0/queries/run/` endpoint.

**Comment by kgodey** [comment]
Moving this to launch blockers; this doesn't block us from usability testing the product.

**Comment by pavish** [comment]
@mathemancer @rajatvijay 

> the root problem is that the front end is changing the returned query at all when saving it. It should never do this

The frontend has to parse the response into its internal modal and then reform the request while saving. It's better to phrase this has to accurately form the same query.

I was not aware we created a separate `distinct_aggregate_to_array`, I assumed we made changes to `aggregate_to_array` function for fixing the distinct issue. We didn't update the frontend for it after these backend changes.

The issue here is that the frontend only knows `aggregate_to_array` and `count`. Since they are the only two options known to the frontend, it there's an unknown option, it selects the first, which is `aggregate_to_array`. We need to change this behaviour to throw an error. I don't think we need to handle this in this as part of this issue itself.

We don't have a UX or usecase for `aggregate_to_array` at the moment, so I think the only major change required in this PR is to change all `aggregate_to_array` functions to `distinct_aggregate_to_array` on the frontend.
